zap/src/main/dist/zap.sh

```diff
@@ -65,7 +65,18 @@ if [ -f "$JVMPROPS" ]; then
   # Local jvm properties file present
   JMEM=$(head -1 "$JVMPROPS")
 elif [ "$OS" = "Linux" ]; then
-  MEM=$(expr $(sed -n 's/MemTotal:[ ]\{1,\}\([0-9]\{1,\}\) kB/\1/p' /proc/meminfo) / 1024)
+    HOSTMEM=$(expr $(sed -n 's/MemTotal:[ ]\{1,\}\([0-9]\{1,\}\) kB/\1/p' /proc/meminfo) / 1024)
+    if [ "$IS_CONTAINERIZED" = "true" ]; then
+        # If we are running in a container, we need to use the cgroup memory limit
+        MEMLIMIT=$(grep hierarchical_memory_limit /sys/fs/cgroup/memory/memory.stat | cut -d' ' -f2)
```

Comment:
Is there a typo here? Checking the ZAP Docker images the file is `/sys/fs/cgroup/memory.stat` (i.e. without the memory dir).

Reply:
@thc202 that's odd, this is a local test:

```
yannickvr@yannickvr:~/zaproxy$ docker run -it softwaresecurityproject/zap-stable /bin/bash
zap@52d82e516a03:/zap$ grep hierarchical_memory_limit /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 9223372036854771712
yannickvr@yannickvr:~/zaproxy$ docker run -m=2048m -it softwaresecurityproject/zap-stable /bin/bash
zap@b56cd3347b54:/zap$ grep hierarchical_memory_limit /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 2147483648
```

Reply:
I guess this is exposed by the OS or something that starts the container. I think it's right that it's not in the container image; it's only available during runtime.

Reply:
Right, I was checking in the running container not the contents of the image, e.g.:
```
➜  ~ docker run -it zaproxy/zap-nightly:latest bash
zap@b5c36f33e445:/zap$ grep hierarchical_memory_limit /sys/fs/cgroup/memory/memory.stat
grep: /sys/fs/cgroup/memory/memory.stat: No such file or directory
```
(Same for `softwaresecurityproject/zap-stable`, also when setting the memory.)

If this depends on the host we should have an additional check that the file really exists (IMO).

Reply:
Agreed. Just out of interest: what OS are you testing on? If we're adding a file check it makes sense to see if we can find how to find the memory usage in your scenario as well

Reply:
Using macOS Sonoma version 14.3.1 (23D60), with Docker version 26.0.0, build 2ae903e.

Reply:
Alright, I'll work on the file check. I don't have access to an Apple system currently. Could you check if this is available?

```
/sys/fs/cgroup/memory/memory.limit_in_bytes
```

Reply:
That one does not exist:
```output
zap@69b5e3df49dc:/zap$ /sys/fs/cgroup/memory/memory.limit_in_bytes
bash: /sys/fs/cgroup/memory/memory.limit_in_bytes: No such file or directory
```

There are these:
```output
zap@69b5e3df49dc:/zap$ ls -la /sys/fs/cgroup/
total 0
drwxr-xr-x  2 root root 0 May  6 16:21 .
drwxr-xr-x 13 root root 0 May  6 16:21 ..
-r--r--r--  1 root root 0 May  6 16:21 cgroup.controllers
-r--r--r--  1 root root 0 May  6 16:21 cgroup.events
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.freeze
--w-------  1 root root 0 May  6 16:21 cgroup.kill
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.max.depth
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.max.descendants
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.procs
-r--r--r--  1 root root 0 May  6 16:21 cgroup.stat
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.subtree_control
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.threads
-rw-r--r--  1 root root 0 May  6 16:21 cgroup.type
-rw-r--r--  1 root root 0 May  6 16:21 cpu.idle
-rw-r--r--  1 root root 0 May  6 16:21 cpu.max
-rw-r--r--  1 root root 0 May  6 16:21 cpu.max.burst
-r--r--r--  1 root root 0 May  6 16:21 cpu.stat
-r--r--r--  1 root root 0 May  6 16:21 cpu.stat.local
-rw-r--r--  1 root root 0 May  6 16:21 cpu.weight
-rw-r--r--  1 root root 0 May  6 16:21 cpu.weight.nice
-rw-r--r--  1 root root 0 May  6 16:21 cpuset.cpus
-r--r--r--  1 root root 0 May  6 16:21 cpuset.cpus.effective
-rw-r--r--  1 root root 0 May  6 16:21 cpuset.cpus.partition
-rw-r--r--  1 root root 0 May  6 16:21 cpuset.mems
-r--r--r--  1 root root 0 May  6 16:21 cpuset.mems.effective
-r--r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.current
-r--r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.events
-r--r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.events.local
-rw-r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.max
-r--r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.numa_stat
-r--r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.rsvd.current
-rw-r--r--  1 root root 0 May  6 16:21 hugetlb.2MB.rsvd.max
-rw-r--r--  1 root root 0 May  6 16:21 io.latency
-rw-r--r--  1 root root 0 May  6 16:21 io.max
-r--r--r--  1 root root 0 May  6 16:21 io.stat
-r--r--r--  1 root root 0 May  6 16:21 memory.current
-r--r--r--  1 root root 0 May  6 16:21 memory.events
-r--r--r--  1 root root 0 May  6 16:21 memory.events.local
-rw-r--r--  1 root root 0 May  6 16:21 memory.high
-rw-r--r--  1 root root 0 May  6 16:21 memory.low
-rw-r--r--  1 root root 0 May  6 16:21 memory.max
-rw-r--r--  1 root root 0 May  6 16:21 memory.min
-rw-r--r--  1 root root 0 May  6 16:21 memory.oom.group
-r--r--r--  1 root root 0 May  6 16:21 memory.peak
--w-------  1 root root 0 May  6 16:21 memory.reclaim
-r--r--r--  1 root root 0 May  6 16:21 memory.stat
-r--r--r--  1 root root 0 May  6 16:21 memory.swap.current
-r--r--r--  1 root root 0 May  6 16:21 memory.swap.events
-rw-r--r--  1 root root 0 May  6 16:21 memory.swap.high
-rw-r--r--  1 root root 0 May  6 16:21 memory.swap.max
-r--r--r--  1 root root 0 May  6 16:21 memory.swap.peak
-r--r--r--  1 root root 0 May  6 16:21 pids.current
-r--r--r--  1 root root 0 May  6 16:21 pids.events
-rw-r--r--  1 root root 0 May  6 16:21 pids.max
-r--r--r--  1 root root 0 May  6 16:21 pids.peak
-r--r--r--  1 root root 0 May  6 16:21 rdma.current
-rw-r--r--  1 root root 0 May  6 16:21 rdma.max
zap@69b5e3df49dc:/zap$ 
```

Reply:
Thanks, that's completely different from anything linux. As memory limits are typically a production thing i've opted for testing if /sys/fs/cgroup/memory/memory.stat exists, that should suffice for now.

Reply:
For the record this seems to be caused by different version of cgroup (i.e. v2 instead of v1), fwiw I'm also seeing this version in Xubuntu 24. Thanks to @ricekot for the details.

Ref: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html